### DISCUSSION TITLE
Restrict asset download to only logged-in users

### DISCUSSION THREAD

**Discussion by iannngit** [statamic/cms#7335] [Q&A] [answered]
Hi all, I'm building a manual download page which allows only logged-in user downloading.

I'm trying a way with route and controller. 
- I registered a route /download
- Change manual file links like /download?file=abc.xyz&container=abc 
- So now it flows to a controller that can get file and container
- I've tried with Statamic\Facade\Asset and Statamic\Facade\AssetContainer to fetch the asset. I have got a Statamic\Assets\Asset object with the correct path as an attribute.
- How can I make it downloadable without redirecting to the path so I don't expose the actual URL and keep the download URL like /download?file= ..... ?

**Comment by iannngit** [comment, marked as answer]
There's an addon for it!  https://github.com/edalzell/statamic-download

**Comment by jasonvarga** [comment]
Yes, you can use Erin's addon. But you were very close anyway. Since you had the `$asset` in a controller, you could have done `return $asset->download();`.
